rosidl_typesupport_introspection_tests/test/test_typesupport_c.cpp

```diff
@@ -0,0 +1,2215 @@
+    action_service_members_structs;
+  std::unordered_map<std::string, const rosidl_typesupport_introspection_c__MessageMembers *>
+    all_message_members_structs;
+  std::unordered_map<std::string, const rosidl_typesupport_introspection_c__ServiceMembers *>
+    all_service_members_structs;
+
+  std::unordered_map<std::string, std::pair<void *, void *>> interfaces_message_constructors = {
+    {"Arrays", {nullptr, nullptr}},
+    {"BasicTypes", {nullptr, nullptr}},
+    {"BoundedPlainSequences", {nullptr, nullptr}},
+    {"BoundedSequences", {nullptr, nullptr}},
+    {"Builtins", {nullptr, nullptr}},
+    {"Constants", {nullptr, nullptr}},
+    {"Defaults", {nullptr, nullptr}},
+    {"Empty", {nullptr, nullptr}},
+    {"MultiNested", {nullptr, nullptr}},
+    {"Nested", {nullptr, nullptr}},
+    {"Strings", {nullptr, nullptr}},
+    {"UnboundedSequences", {nullptr, nullptr}},
+    {"WStrings", {nullptr, nullptr}},
+  };
+  std::unordered_map<std::string, std::pair<void *, void *>> service_message_constructors = {
+    {"Arrays_Request", {nullptr, nullptr}},
+    {"Arrays_Response", {nullptr, nullptr}},
+    {"BasicTypes_Request", {nullptr, nullptr}},
+    {"BasicTypes_Response", {nullptr, nullptr}},
+    {"Empty_Request", {nullptr, nullptr}},
+    {"Empty_Response", {nullptr, nullptr}},
+  };
+  std::unordered_map<std::string, std::pair<void *, void *>> action_message_constructors = {
+    {"Fibonacci_Feedback", {nullptr, nullptr}},
+    {"Fibonacci_FeedbackMessage", {nullptr, nullptr}},
+    {"Fibonacci_GetResult_Request", {nullptr, nullptr}},
+    {"Fibonacci_GetResult_Response", {nullptr, nullptr}},
+    {"Fibonacci_Goal", {nullptr, nullptr}},
+    {"Fibonacci_Result", {nullptr, nullptr}},
+    {"Fibonacci_SendGoal_Request", {nullptr, nullptr}},
+    {"Fibonacci_SendGoal_Response", {nullptr, nullptr}},
+    {"NestedMessage_Feedback", {nullptr, nullptr}},
+    {"NestedMessage_FeedbackMessage", {nullptr, nullptr}},
+    {"NestedMessage_GetResult_Request", {nullptr, nullptr}},
+    {"NestedMessage_GetResult_Response", {nullptr, nullptr}},
+    {"NestedMessage_Goal", {nullptr, nullptr}},
+    {"NestedMessage_Result", {nullptr, nullptr}},
+    {"NestedMessage_SendGoal_Request", {nullptr, nullptr}},
+    {"NestedMessage_SendGoal_Response", {nullptr, nullptr}},
+  };
+  std::unordered_map<std::string, std::pair<void *, void *>> all_message_constructors;
+
+  void * types_lib = nullptr;
+
+  std::string make_message_create_function_name(
+    std::string ns,
+    std::string type,
+    std::string interface)
+  {
+    return ns + "__" + type + "__" + interface + "__destroy";
+  }
+
+  std::string make_message_destroy_function_name(
+    std::string ns,
+    std::string type,
+    std::string interface)
+  {
+    return ns + "__" + type + "__" + interface + "__destroy";
+  }
+
+  void populate_constructors()
+  {
+    auto typesupport_lib_name = "lib" + ns + "__rosidl_generator_c.so";
```

Comment:
@gbiggs meta: did you consider templating the hell of out this instead of dynamic loading interface representations (i.e. `__rosidl_generator_c*.so` libraries)? Few tests below benefit from the latter. And I'm a bit wary about name mangling when it's C++ and not C that we're loading.

I'm thinking that some type traits and fully specialized fixtures could come in handy.

Reply:
Actually, where are these constructors being used?

Reply:
I don't think they are being used. This is left-over code from an avenue I was exploring for testing the message introspection. I forgot to remove these functions and variables when I abandoned that idea.